doc/in-progress/object_versions.md

```diff
@@ -0,0 +1,146 @@
+                    id: object::new(ctx),
+                    x: 42,
+                }
+            }
+            tx_context::sender(ctx),
+        )
+    }
+}
+```
+
+You can't access wrapped objects by their ID in the object store, you must access them by the object that wraps them. In the prior example, the owner of `Outer` must specify it as the transaction input and then access its `inner` field to read the instance of `Inner`. Validators refuse to sign transactions that specify wrapped objects (like the `inner` of an `Outer`) as inputs.  As a result, you don't need to specify a wrapped object's version in a transaction that reads that object.
+
+Wrapped objects can eventually become "unwrapped", meaning that they are once again accessible at their ID:
+
+```
+module example::wrapped {
+    // ...
+    
+    entry fun unwrap(outer: Outer, ctx: &TxContext) {
+        let Outer { id, inner } = outer;
+        object::delete(id);
+        transfer::transfer(inner, tx_context::sender(ctx));
+    }
+}
+```
+
+The `unwrap` function in the previous code takes an instance of `Outer`, destroys it, and sends the `Inner` back to the sender.  After calling this function, the previous owner of `Outer` can access `Inner` directly by its ID because it is now unwrapped.  Wrapping and unwrapping of an object can happen multiple times across its lifespan, and the object retains its ID across all those events.
+
+The Lamport timestamp-based versioning scheme ensures that the version that an object is unwrapped at is always greater than the version it was wrapped at, to prevent version re-use.
+
+- After a transaction, `W` wrapping object `I` in object `O`, `O`'s version is greater than or equal to `I`'s. This means one of the following conditions is true:
+  - `I` is an input so has a strictly lower version.
+  - `I` is new and has an equal version.
+- After a later transaction unwrapping `I` out of `O`, the following must be true: 
+  - `O`'s input version is greater than or equal to its version after `W` because it is a later transaction, so the version can only have increased.
+  - `I`'s version in the output must be strictly greater than `O`'s input version.
+  
+This leads to the following chain of inequalities for `I`'s version before wrapping:
+
+- less than or equal to `O`'s version after wrapping
+- less than or equal to `O`'s version before unwrapping
+- less than `I`'s version after unwrapping
+
+So `I`'s version before wrapping is less than `I`'s version after unwrapping.
+
+### Dynamic Fields
+
+From a versioning perspective, values held in dynamic fields behave like wrapped objects:
+
+- They are only accessible via their field owner, not as direct transaction inputs.
+- Based on the previous point, you do not need to supply their versions with the transaction inputs.
+- Lamport timestamp-based versioning makes sure that when a transaction removes a field and its value becomes accessible by its ID, its version has been incremented.
+
```

Comment:
Big thing missing here is that Lamport versioning solves the creation/deletion/creation problem. I tried writing a short blurb about it but its a bit of a mess
```
In order to truly understand what problem Lamport versioning solves for dynamic fields, we need to go a bit more in depth into how dynamic fields are implemented. Dynamic fields are objects themselves, but the owner of the field object is another object. We call this object the parent object. When creating the dynamic field, it's ID is _not_ created based on the Transaction Digest. Instead we hash information that is used to access the field, specifically the ID of the parent, the "name" of the field (a Move value), and the type of the "name" (the type of the Move value). In other words, the ID of a dynamic field object is determined by `hash(parent.id || name || type(name))`. 
The difficulty with versioning is that the dynamic field object can be deleted and then in a later transaction recreated! But as we stated above, the version for an object cannot be reused. This creates a problem when we delete a dynamic field and then recreate it: how do we find a version that is greater than all previous versions? 
Lamport versioning gives us exactly that. Since the dynamic field can only be accessed through its parent, we know that when it was deleted its version was less than or equal to its parent's version. And when it is recreated, we will give it the Lamport version for the transaction. Since that version is increasing for the parent, it must also be greater than any perviously used version for the dynamic field object. 

(Note: there is some subtlety here about dynamic fields on wrapped objects, but the logic works the same. The object containing the wrapped parent can effectively be viewed as the parent of the dynamic field, at least as far as versioning goes)
```